File: euler38.py
# Какое самое большое девятизначное пан-цифровое число можно образовать
# как объединенное произведение целого числа и (1,2, ... , n), где n > 1?

from project_euler_defs import *

n = 10

# 1 вариант
start_time = datetime.now()
pan_digits = []
template = [str(_) for _ in range(1, n)]
for y in range(n * 1000):
    pan_digit_check, x = str(y), 2
    while len(pan_digit_check + str(y * x)) < n:
        pan_digit_check += str(y * x)
        x += 1
    if sorted(pan_digit_check) == template:
        pan_digits.append(pan_digit_check)
print(max(pan_digits))
print(datetime.now() - start_time)

# 2 вариант
start_time = datetime.now()
pass
print(datetime.now() - start_time)

File: go/euler38.go
// euler38 находит самое большое девятизначное пан-цифровое число.
//
// Решение задачи Эйлера №38.
package main

import (
	"fmt"
	"sort"
	"time"
)

func euler38() {
	n := 10

	startTime := time.Now()
	panDigits := make([]string, 0)
	template := make([]string, n-1)
	for i := 1; i < n; i++ {
		template[i-1] = fmt.Sprintf("%d", i)
	}
	sort.Strings(template)
	for y := 0; y < n*1000; y++ {
		panDigitCheck := fmt.Sprintf("%d", y)
		x := 2
		for len(panDigitCheck)+len(fmt.Sprintf("%d", y*x)) < n {
			panDigitCheck += fmt.Sprintf("%d", y*x)
			x++
		}
		check := make([]string, 0)
		for _, c := range panDigitCheck {
			check = append(check, string(c))
		}
		if sort.StringsAreSorted(check) == (len(check) > 0 && check[0] < check[len(check)-1]) {
			// Simplified check - just add if length matches
			if len(panDigitCheck) == n-1 {
				isPan := true
				seen := make(map[string]bool)
				for _, c := range panDigitCheck {
					s := string(c)
					if seen[s] {
						isPan = false
						break
					}
					seen[s] = true
				}
				if isPan {
					panDigits = append(panDigits, panDigitCheck)
				}
			}
		}
	}
	maxPan := ""
	for _, p := range panDigits {
		if p > maxPan {
			maxPan = p
		}
	}
	fmt.Println(maxPan)
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	fmt.Println(time.Since(startTime))
}

func main() {
	euler38()
}
